src/lib/branchpredictors/PerceptronPredictor.cc

```diff
@@ -116,11 +145,14 @@ void PerceptronPredictor::update(uint64_t address, bool taken,
   }
 
   btb_[hashedIndex].first = perceptron;
-  btb_[hashedIndex].second = targetAddress;
+  if (isTaken) {
+    btb_[hashedIndex].second = targetAddress;
+  }
 
-  globalHistory_ =
-      ((globalHistory_ << 1) | taken) & ((1 << globalHistoryLength_) - 1);
-  return;
+  // Update global history if prediction was incorrect
+  // Bit-flip the global history bit corresponding to this prediction
+  // We know how many predictions there have since been by the size of the FTQ
+  if (directionPrediction != isTaken) globalHistory_ ^= (1 << (ftq_.size()));
 }
 
 void PerceptronPredictor::flush(uint64_t address) {
```

Comment:
In a similar way to update, I think (tell me if I'm wrong) that flush is **required** to be called in reverse? program order due to the way you are using `pop_back()` etc. If so I think there should be an assert the same as in update to enforce this. (applicable to any BP with a ras or ftq)

Reply:
Nope, doesn't need to be flushed in reverse order.

Reply:
Just found the comment in `BranchPredictor.hh:35`